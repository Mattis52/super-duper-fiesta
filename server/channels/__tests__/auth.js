jest.mock('../../models/meeting');
jest.mock('../../models/user');
const { register } = require('../auth');
const { getActiveGenfors } = require('../../models/meeting');
const { getAnonymousUser, addAnonymousUser, getUserByUsername } = require('../../models/user');
const { generateGenfors, generateSocket, generateAnonymousUser, generateUser } = require('../../utils/generateTestData');

beforeEach(() => {
  getActiveGenfors.mockImplementation(async () => generateGenfors({ pin: 1234567890 }));
  getAnonymousUser.mockImplementation(async () => null);
  addAnonymousUser.mockImplementation(
    async () => {},
  );
  getUserByUsername.mockImplementation(async (username, genfors) =>
    generateUser({ username, genfors, completedRegistration: false }));
});

const generateData = data => (Object.assign({
  pin: 1234567890,
  passwordHash: null,
}, data));


describe('register', () => {
  it('emits registration status', async () => {
    const socket = generateSocket({ completedRegistration: false });
    await register(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
    expect(socket.broadcast.emit.mock.calls).toMatchSnapshot();
  });

  it('emits error when registration is closed', async () => {
    getActiveGenfors.mockImplementation(async () => generateGenfors({ registrationOpen: false }));
    const socket = generateSocket({ completedRegistration: false });
    await register(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });

  it('emits error when pin code is wrong', async () => {
    getActiveGenfors.mockImplementation(async () => generateGenfors({ pin: 5453577654 }));
    const socket = generateSocket({ completedRegistration: false });
    await register(socket, generateData());

    expect(socket.emit.mock.calls).toMatchSnapshot();
    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });


  it('emits error when user is already registered with wrong personal code', async () => {
    const socket = generateSocket({ completedRegistration: true }, { passwordHash: 'correct' });
    await register(
      socket,
      generateData({ passwordHash: 'wrong' }),
    );

    expect(socket.emit.mock.calls).toMatchSnapshot();
    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });


  it('emits registered when user is already registered with correct personal code', async () => {
    getAnonymousUser.mockImplementation(
      async (passwordHash, onlinewebId, genfors) => generateAnonymousUser({
        passwordHash,
        genfors,
      },
    ));
    const socket = generateSocket({ completedRegistration: true }, { passwordHash: 'correct' });
    await register(
      socket,
      generateData({ passwordHash: 'correct' }),
    );

    expect(socket.emit.mock.calls).toMatchSnapshot();
    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });

  it('emits error when handling errors when validating hash', async () => {
    getAnonymousUser.mockImplementation(
      async () => {
        throw new Error('Failed');
      },
    );
    const socket = generateSocket({ completedRegistration: true }, { passwordHash: 'correct' });
    await register(
      socket,
      generateData({ passwordHash: 'correct' }),
    );

    expect(socket.emit.mock.calls).toMatchSnapshot();
    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });

  it('emits error when handling errors when saving anonymous user', async () => {
    addAnonymousUser.mockImplementation(
      async () => {
        throw new Error('Failed');
      },
    );
    const socket = generateSocket({ completedRegistration: false });
    await register(
      socket,
      generateData(),
    );

    expect(socket.emit.mock.calls).toMatchSnapshot();
    expect(socket.broadcast.emit.mock.calls).toEqual([]);
  });
});
