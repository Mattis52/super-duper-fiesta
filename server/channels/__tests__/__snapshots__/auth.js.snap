// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register emits error when handling errors when saving anonymous user 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "AUTH_ERROR",
    Object {
      "error": "Noe gikk galt under registreringen. Prøv igjen",
    },
  ],
]
`;

exports[`register emits error when handling errors when validating hash 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "correct",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "AUTH_ERROR",
    Object {
      "error": "Validering av personlig kode feilet",
    },
  ],
]
`;

exports[`register emits error when pin code is wrong 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "AUTH_ERROR",
    Object {
      "error": "Feil pinkode",
    },
  ],
]
`;

exports[`register emits error when registration is closed 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "AUTH_ERROR",
    Object {
      "error": "Registreringen er ikke åpen.",
    },
  ],
]
`;

exports[`register emits error when user is already registered with wrong personal code 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "correct",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "AUTH_ERROR",
    Object {
      "error": "Feil personlig kode",
    },
  ],
]
`;

exports[`register emits registered when user is already registered with correct personal code 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "correct",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": true,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "AUTH_REGISTERED",
    Object {
      "registered": true,
    },
  ],
]
`;

exports[`register emits registration status 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "headers": Object {
          "cookie": Object {
            "passwordHash": "hashy",
          },
        },
        "user": Object {
          "_id": "123",
          "completedRegistration": false,
          "genfors": "1",
          "logged_in": true,
          "name": "Namy",
          "onlinewebId": "123",
          "permissions": 5,
        },
      },
    },
    "AUTH_REGISTERED",
    Object {
      "registered": true,
    },
  ],
]
`;
